### PR TITLE
Remove most RemoveAsAliasesFromSql usages

### PR DESCRIPTION
**Pull request by bahusoid**
As CriterionUtil handles better property projections and it's one step further to properly support multi-column projections.

Added optimized versions of CriterionUtil.GetColumn methods that don't create SqlString for property projections

Related to #846

### CODE REVIEW

File: src/NHibernate/Criterion/GroupedProjection.cs [hunk 1]
```diff
@@ -18,13 +17,12 @@ public GroupedProjection(IProjection projection)
...
 		{
-			return renderedProjection = projection.ToSqlString(criteria, position, criteriaQuery);
+			return projection.ToSqlString(criteria, position, criteriaQuery);
 		}
 
 		public virtual SqlString ToGroupSqlString(ICriteria criteria, ICriteriaQuery criteriaQuery)
 		{
-			//This is kind of a hack. The hack is based on the fact that ToGroupSqlString always called after ToSqlString.
-			return SqlStringHelper.RemoveAsAliasesFromSql(renderedProjection);
+			return SqlStringHelper.Join(new SqlString(","), CriterionUtil.GetColumnNamesAsSqlStringParts(null, projection, criteriaQuery, criteria));
```
Comment (hazzik) [inline src/NHibernate/Criterion/GroupedProjection.cs]: This would break.

Comment (hazzik) [reply src/NHibernate/Criterion/GroupedProjection.cs]:
```
3) Failed : NHibernate.Test.NHSpecificTest.NH3609.Fixture.GroupByClauseHasParameterSet
  Expected: No Exception to be thrown
  But was:  <NHibernate.Exceptions.GenericADOException: could not execute query
[ SELECT (case when this_.Name = @p0 then @p1 else @p2 end) as y0_ FROM Entity this_ GROUP BY (case when this_.Name = @p3 then @p4 else @p5 end) ]
  Name:cp0 - Value:  Name:cp1 - Value:1  Name:cp2 - Value:2  Name:cp3 - Value:  Name:cp4 - Value:1  Name:cp5 - Value:2
[SQL: SELECT (case when this_.Name = @p0 then @p1 else @p2 end) as y0_ FROM Entity this_ GROUP BY (case when this_.Name = @p3 then @p4 else @p5 end)] ---> System.Data.SqlClient.SqlException: Column 'Entity.Name' is invalid in the select list because it is not contained in either an aggregate function or the GROUP BY clause.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at NHibernate.AdoNet.AbstractBatcher.ExecuteReader(DbCommand cmd) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/AdoNet/AbstractBatcher.cs:line 226
   at NHibernate.Loader.Loader.GetResultSet(DbCommand st, QueryParameters queryParameters, ISessionImplementor session, IResultTransformer forcedResultTransformer) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 1516
   at NHibernate.Loader.Loader.DoQuery(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies, IResultTransformer forcedResultTransformer, QueryCacheResultBuilder queryCacheResultBuilder) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 487
   at NHibernate.Loader.Loader.DoQueryAndInitializeNonLazyCollections(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies, IResultTransformer forcedResultTransformer, QueryCacheResultBuilder queryCacheResultBuilder) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 285
   at NHibernate.Loader.Loader.DoList(ISessionImplementor session, QueryParameters queryParameters, IResultTransformer forcedResultTransformer, QueryCacheResultBuilder queryCacheResultBuilder) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 1956
   --- End of inner exception stack trace ---
   at NHibernate.Loader.Loader.DoList(ISessionImplementor session, QueryParameters queryParameters, IResultTransformer forcedResultTransformer, QueryCacheResultBuilder queryCacheResultBuilder) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 1965
   at NHibernate.Impl.SessionImpl.List(CriteriaImpl criteria, IList results) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Impl/SessionImpl.cs:line 1734
   at NHibernate.Impl.CriteriaImpl.List(IList results) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Impl/CriteriaImpl.cs:line 282
   at NHibernate.Impl.CriteriaImpl.List[T]() in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Impl/CriteriaImpl.cs:line 294
   at NHibernate.Test.NHSpecificTest.NH3609.Fixture.<>c__DisplayClass4_1.<GroupByClauseHasParameterSet>b__0() in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate.Test/NHSpecificTest/NH3609/Fixture.cs:line 100
--- End of stack trace from previous location where exception was thrown ---
   at NUnit.Framework.Internal.Reflect.DynamicInvokeWithTransparentExceptions(Delegate delegate)
   at NUnit.Framework.Internal.ExceptionHelper.RecordException(Delegate parameterlessDelegate, String parameterName)>
   at NHibernate.Test.NHSpecificTest.NH3609.Fixture.GroupByClauseHasParameterSet() in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate.Test/NHSpecificTest/NH3609/Fixture.cs:line 98
```

Comment (bahusoid) [reply src/NHibernate/Criterion/GroupedProjection.cs]: Yeah - `RemoveAsAliasesFromSql` can't be removed here that easily. I restored it and added logic to handle `IPropertyProjection`.

File: src/NHibernate/Criterion/PropertyExpression.cs [hunk 2]
```diff
@@ -70,25 +70,25 @@ protected PropertyExpression(string lhsPropertyName, IProjection rhsProjection)
...
+				CriterionUtil.GetColumnNamesAsSqlStringParts(_rhsPropertyName, _rhsProjection, criteriaQuery, criteria);
 
 			SqlStringBuilder sb = new SqlStringBuilder();
 			if (columnNames.Length > 1)
 			{
 				sb.Add(StringHelper.OpenParen);
 			}
 			bool first = true;
-			foreach (SqlString sqlString in SqlStringHelper.Add(columnNames, Op, otherColumnNames))
+			for (var i = 0; i < columnNames.Length; i++)
```
Comment (fredericDelaporte) [inline src/NHibernate/Criterion/PropertyExpression.cs:84]: Since this loop is rewritten anyway, why not use the `if (columnNames.Length > 1)` above to handle the first case then iterate from `1` and remove the `first` bloat?

File: src/NHibernate/Criterion/PropertyExpression.cs [hunk 3]
```diff
@@ -70,25 +70,25 @@ protected PropertyExpression(string lhsPropertyName, IProjection rhsProjection)
...
 			{
 				if (first == false)
 				{
 					sb.Add(" and ");
 				}
 				first = false;
```
Comment (fredericDelaporte) [inline src/NHibernate/Criterion/PropertyExpression.cs:90]: Or switch that to:
```c#
if (first)
    first = false;
else
    sb.Add(" and ");
```

File: src/NHibernate/SqlCommand/SqlStringHelper.cs [hunk 4]
```diff
@@ -43,6 +43,7 @@ public static SqlString[] Add(SqlString[] x, string sep, SqlString[] y)
 		}
 
 
+		//Consider using CriterionUtil.GetColumn... methods instead
```
Comment (fredericDelaporte) [inline src/NHibernate/SqlCommand/SqlStringHelper.cs:46]: Consider removing the above undue blank line by the way.
And also, for better discoverability of the advice:
```suggestion
		/// <summary>
		/// Remove the <c>as someColumnAlias<c/> clause from a <c>SqlString</c> representing a column expression.
		/// Consider using <c>CriterionUtil.GetColumn...</c> methods instead.
		/// </summary>
```

Comment (bahusoid) [reply src/NHibernate/SqlCommand/SqlStringHelper.cs]: Done

File: src/NHibernate/Criterion/PropertyExpression.cs (hunk 3 above)
Comment (bahusoid) [reply src/NHibernate/Criterion/PropertyExpression.cs]: `first` is removed

**Comment by fredericDelaporte** [comment]
So this is a partial fix of #846.

File: src/NHibernate/Criterion/InExpression.cs [hunk 5]
```diff
@@ -64,24 +64,22 @@ public override SqlString ToSqlString(ICriteria criteria, ICriteriaQuery criteri
 
 			//TODO: add default capacity
 			SqlStringBuilder result = new SqlStringBuilder();
-			SqlString[] columnNames =
-				CriterionUtil.GetColumnNames(_propertyName, _projection, criteriaQuery, criteria);
+			var columnNames =
+				CriterionUtil.GetColumnNamesAsSqlStringParts(_propertyName, _projection, criteriaQuery, criteria);
```
Comment (fredericDelaporte) [inline src/NHibernate/Criterion/InExpression.cs:68]: Changes in this file conflict with #2158. To me it seems those changes should be dropped. I let you see by yourself.

Comment (bahusoid) [reply src/NHibernate/Criterion/InExpression.cs]: Done. Yeah just dropped them

File: src/NHibernate/Criterion/PropertyExpression.cs (hunk 2 above)
Comment (hazzik) [reply src/NHibernate/Criterion/PropertyExpression.cs:89]: Done

File: src/NHibernate/Criterion/PropertyExpression.cs [hunk 6]
```diff
@@ -70,31 +70,25 @@ protected PropertyExpression(string lhsPropertyName, IProjection rhsProjection)
...
-			foreach (SqlString sqlString in SqlStringHelper.Add(columnNames, Op, otherColumnNames))
-			{
-				if (first == false)
-				{
-					sb.Add(" and ");
-				}
-				first = false;
-				sb.Add(sqlString);
-			}
+			if (columnNames.Length == 0)
```
Comment (bahusoid) [inline src/NHibernate/Criterion/PropertyExpression.cs:78]: Why did you change order of checks? In most cases it's 1 so it was first check intentionally. And it should never be 0 - it's added just for consistency with the old logic.
